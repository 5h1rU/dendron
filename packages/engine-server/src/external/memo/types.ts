import { URI } from "vscode-uri";

export type WorkspaceCache = {
  imageUris: URI[];
  markdownUris: URI[];
  otherUris: URI[];
  allUris: URI[];
  danglingRefs: string[];
  danglingRefsByFsPath: { [key: string]: string[] };
};

export type RefT = {
  label: string;
  ref: string;
};
