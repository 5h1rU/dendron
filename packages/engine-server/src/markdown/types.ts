import { DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
  children: MDAST.Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  // MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
};

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};
