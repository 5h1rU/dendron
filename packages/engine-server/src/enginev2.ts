import {
  ConfigWriteOpts,
  DendronConfig,
  DendronError,
  DEngineClientV2,
  DEngineDeleteSchemaRespV2,
  DEngineInitRespV2,
  DEngineMode,
  DEngineV2,
  DLink,
  DNodeTypeV2,
  DStoreV2,
  DVault,
  EngineDeleteOptsV2,
  EngineUpdateNodesOptsV2,
  EngineWriteOptsV2,
  GetNoteOptsV2,
  GetNotePayloadV2,
  NoteChangeEntry,
  NotePropsDictV2,
  NotePropsV2,
  NoteUtilsV2,
  QueryNotesOpts,
  RenameNoteOptsV2,
  RenameNotePayload,
  RespV2,
  SchemaModuleDictV2,
  SchemaModulePropsV2,
  SchemaQueryResp,
  WorkspaceOpts,
  WriteNoteResp,
} from "@dendronhq/common-all";
import {
  createLogger,
  DLogger,
  readYAML,
  VaultUtils,
  writeYAML,
} from "@dendronhq/common-server";
import _ from "lodash";
import { DConfig } from "./config";
import { FileStorageV2 } from "./drivers/file/storev2";
import { FuseEngine } from "./fuseEngine";

type DendronEngineOptsV2 = {
  wsRoot: string;
  vaultsv3: DVault[];
  forceNew?: boolean;
  store?: any;
  mode?: DEngineMode;
  logger?: DLogger;
};
type DendronEnginePropsV2 = Required<DendronEngineOptsV2>;

export class DendronEngineV2 implements DEngineV2 {
  public wsRoot: string;
  public vaults: string[];
  public store: DStoreV2;
  protected props: DendronEnginePropsV2;
  public logger: DLogger;
  public fuseEngine: FuseEngine;
  public links: DLink[];
  public vaultsv3: DVault[];
  public configRoot: string;

  static _instance: DendronEngineV2 | undefined;

  constructor(props: DendronEnginePropsV2) {
    this.wsRoot = props.wsRoot;
    this.configRoot = props.wsRoot;
    this.store = props.store;
    this.logger = props.logger;
    this.props = props;
    this.fuseEngine = new FuseEngine({});
    this.links = [];
    this.vaultsv3 = props.vaultsv3;
    this.vaults = props.vaultsv3.map((ent) => ent.fsPath);
  }

  static createV3({
    vaults,
    wsRoot,
    logger,
  }: WorkspaceOpts & { logger?: DLogger }) {
    const LOGGER = logger || createLogger();
    return new DendronEngineV2({
      wsRoot,
      vaultsv3: vaults,
      forceNew: true,
      store: new FileStorageV2({
        wsRoot,
        vaultsv3: vaults,
        logger: LOGGER,
      }),
      mode: "fuzzy",
      logger: LOGGER,
    });
  }

  static instance({ vaults, wsRoot }: { vaults: DVault[]; wsRoot: string }) {
    if (!DendronEngineV2._instance) {
      DendronEngineV2._instance = DendronEngineV2.createV3({ vaults, wsRoot });
    }
    return DendronEngineV2._instance;
  }

  get notes(): NotePropsDictV2 {
    return this.store.notes;
  }
  get schemas(): SchemaModuleDictV2 {
    return this.store.schemas;
  }

  /**
   * Does not throw error but returns it
   */
  async init(): Promise<DEngineInitRespV2> {
    try {
      const { data, error } = await this.store.init();
      const { notes, schemas } = data;
      this.updateIndex("note");
      this.updateIndex("schema");
      return {
        error,
        data: { notes, schemas },
      };
    } catch (error) {
      const { message, stack, msg, status, friendly } = error;
      let payload = { message, stack };
      return {
        error: new DendronError({ payload, msg, status, friendly }),
        data: {
          notes: {},
          schemas: {},
        },
      };
    }
  }

  async deleteNote(
    id: string,
    opts?: EngineDeleteOptsV2
  ): ReturnType<DEngineClientV2["deleteNote"]> {
    try {
      const note = this.notes[id];
      const changed = await this.store.deleteNote(id, opts);
      const noteChangeEntry = _.find(
        changed,
        (ent) => ent.note.id === id
      ) as NoteChangeEntry;
      if (noteChangeEntry.status === "delete") {
        await this.fuseEngine.removeNoteFromIndex(note);
      }
      return {
        data: changed,
        error: null,
      };
    } catch (err) {
      return {
        data: [],
        error: err,
      };
    }
  }

  async deleteSchema(
    id: string,
    opts?: EngineDeleteOptsV2
  ): Promise<DEngineDeleteSchemaRespV2> {
    try {
      const data = (await this.store.deleteSchema(
        id,
        opts
      )) as DEngineDeleteSchemaRespV2;
      // deleted schema might affect notes
      await this.updateIndex("note");
      await this.updateIndex("schema");
      return data;
      // FIXM:E not performant
      // const smod = this.schemas[id];
      // await this.fuseEngine.removeSchemaFromIndex(smod);
      // return {
      //   data: undefined,
      //   error: null,
      // };
    } catch (err) {
      return {
        data: {
          notes: {},
          schemas: {},
        },
        error: err,
      };
    }
  }

  async getNoteByPath({
    npath,
    createIfNew,
    vault,
  }: GetNoteOptsV2): Promise<RespV2<GetNotePayloadV2>> {
    const ctx = "getNoteByPath";
    this.logger.debug({ ctx, npath, createIfNew, msg: "enter" });
    const maybeNote = NoteUtilsV2.getNoteByFname(npath, this.notes, {
      vault: vault,
    });
    this.logger.debug({ ctx, maybeNote, msg: "post-query" });
    let noteNew: NotePropsV2 | undefined = maybeNote;
    let changed: NoteChangeEntry[] = [];
    let error = null;
    if ((!maybeNote || maybeNote.stub) && createIfNew) {
      this.logger.debug({ ctx, maybeNote, msg: "create-new" });
      if (maybeNote?.stub) {
        noteNew = maybeNote;
        delete noteNew.stub;
      } else {
        noteNew = NoteUtilsV2.createWithSchema({
          noteOpts: { fname: npath, vault },
          engine: this,
        });
      }
      changed = (await this.writeNote(noteNew, { newNode: true })).data;
    }
    if (!createIfNew && !maybeNote) {
      error = new DendronError({ status: "no_note_found" });
    }
    return {
      data: { note: noteNew, changed },
      error,
    };
  }

  async getConfig() {
    const cpath = DConfig.configPath(this.configRoot);
    const config = readYAML(cpath) as DendronConfig;
    return {
      error: null,
      data: config,
    };
  }

  async getSchema(id: string): Promise<RespV2<SchemaModulePropsV2>> {
    const ctx = "getSchema";
    const data = this.schemas[id];
    this.logger.info({ ctx, msg: "exit" });
    return {
      data,
      error: null,
    };
  }

  queryNotesSync({
    qs,
  }: {
    qs: string;
  }): ReturnType<DEngineClientV2["queryNotesSync"]> {
    const items = this.fuseEngine.queryNote({ qs });
    return {
      error: null,
      data: items.map((ent) => this.notes[ent.id]),
    };
  }

  async querySchema(queryString: string): Promise<SchemaQueryResp> {
    const ctx = "querySchema";

    let items: SchemaModulePropsV2[] = [];
    const results = await this.fuseEngine.querySchema({ qs: queryString });
    items = results.map((ent) => this.schemas[ent.id]);
    // if (queryString === "") {
    //   items = [this.schemas.root];
    // } else if (queryString === "*") {
    //   items = _.values(this.schemas);
    // } else {
    //   const results = this.schemaIndex.search(queryString);
    //   items = _.map(results, (resp) => this.schemas[resp.item.id]);
    // }
    this.logger.info({ ctx, msg: "exit" });
    return {
      error: null,
      data: items,
    };
  }

  async queryNotes(
    opts: QueryNotesOpts
  ): ReturnType<DEngineClientV2["queryNotes"]> {
    const ctx = "Engine:queryNotes";
    const { qs, vault, createIfNew } = opts;
    let items = await this.fuseEngine.queryNote({ qs });
    let item = this.notes[items[0].id];
    if (createIfNew) {
      let noteNew: NotePropsV2;
      if (item?.fname === qs && item?.stub) {
        noteNew = item;
        noteNew.stub = false;
      } else {
        if (_.isUndefined(vault)) {
          return {
            error: new DendronError({ msg: "no vault specified" }),
            data: null as any,
          };
        }
        noteNew = NoteUtilsV2.create({ fname: qs, vault });
      }
      await this.writeNote(noteNew, { newNode: true });
    }
    this.logger.info({ ctx, msg: "exit" });
    let notes = items.map((ent) => this.notes[ent.id]);
    if (!_.isUndefined(vault)) {
      notes = notes.filter((ent) => VaultUtils.isEqual(vault, ent.vault));
    }
    return {
      error: null,
      data: notes,
    };
  }

  // async query(
  //   queryString: string,
  //   mode: DNodeTypeV2,
  //   opts?: QueryOptsV2
  // ): Promise<EngineQueryNoteResp> {
  //   const ctx = "Engine:query";
  //   const cleanOpts = _.defaults(opts || {}, {
  //     fullNode: false,
  //     createIfNew: false,
  //     initialQuery: false,
  //     stub: false,
  //   });
  //   this.logger.info({ ctx, msg: "enter" });
  //   let items: DNodePropsV2[] = [];

  //   // ~~~ schema query
  //   if (mode === "schema") {
  //     throw Error("engine.query for schema is not supported");
  //   } else {
  //     // ~~~ note query
  //     items = await this.fuseEngine.queryNote({ qs: queryString });
  //     // if (queryString === "") {
  //     //   items = [this.notes.root];
  //     // } else if (queryString === "*") {
  //     //   items = _.values(this.notes);
  //     // } else {
  //     //   const results = this.notesIndex.search(queryString);
  //     //   items = _.map(results, (resp) => resp.item);
  //     // }
  //     if (cleanOpts.createIfNew) {
  //       let noteNew: NotePropsV2;
  //       if (items[0]?.fname === queryString && items[0]?.stub) {
  //         noteNew = items[0];
  //         noteNew.stub = false;
  //       } else {
  //         noteNew = NoteUtilsV2.create({ fname: queryString });
  //       }
  //       await this.writeNote(noteNew, { newNode: true });
  //     }
  //     if (cleanOpts.fullNode) {
  //       throw Error("not implemented");
  //     }
  //   }

  //   // ~~~ exit
  //   this.logger.info({ ctx, msg: "exit" });
  //   return {
  //     error: null,
  //     data: items,
  //   };
  // }

  async sync() {
    throw Error("sync not implemented");
    return {} as any;
  }

  async renameNote(opts: RenameNoteOptsV2): Promise<RespV2<RenameNotePayload>> {
    const resp = await this.store.renameNote(opts);
    return {
      error: null,
      data: resp,
    };
  }

  async updateNote(
    note: NotePropsV2,
    opts?: EngineUpdateNodesOptsV2
  ): Promise<void> {
    const out = this.store.updateNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async updateIndex(mode: DNodeTypeV2) {
    if (mode === "schema") {
      this.fuseEngine.updateSchemaIndex(this.schemas);
    } else {
      this.fuseEngine.updateNotesIndex(this.notes);
    }
  }

  async updateSchema(schemaModule: SchemaModulePropsV2) {
    const out = await this.store.updateSchema(schemaModule);
    await this.updateIndex("schema");
    return out;
  }

  async writeConfig(
    opts: ConfigWriteOpts
  ): ReturnType<DEngineV2["writeConfig"]> {
    const { configRoot } = this;
    const cpath = DConfig.configPath(configRoot);
    writeYAML(cpath, opts.config);
    return {
      error: null,
    };
  }

  async writeNote(
    note: NotePropsV2,
    opts?: EngineWriteOptsV2
  ): Promise<WriteNoteResp> {
    const out = await this.store.writeNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async writeSchema(schema: SchemaModulePropsV2) {
    return this.store.writeSchema(schema);
  }
}
