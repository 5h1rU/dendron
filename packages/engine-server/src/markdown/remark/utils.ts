import {
  CONSTANTS,
  DendronError,
  getSlugger,
  NoteChangeEntry,
  NoteProps,
  NoteUtils,
} from "@dendronhq/common-all";
import { normalizev2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { Heading, Root } from "mdast";
import { Processor } from "unified";
import { Node } from "unist";
import { selectAll } from "unist-util-select";
import { VFile } from "vfile";
import { WikiLinkProps } from "../../topics/markdown";
import {
  DendronASTRoot,
  DendronASTTypes,
  NoteRefNoteV4_LEGACY,
} from "../types";

export const ALIAS_DIVIDER = "|";

export function addError(proc: Processor, err: DendronError) {
  const errors = proc.data("errors") as DendronError[];
  errors.push(err);
  proc().data("errors", errors);
}

export function getNoteOrError(
  notes: NoteProps[],
  hint: any
): { error: DendronError | undefined; note: undefined | NoteProps } {
  let error: DendronError | undefined;
  let note: NoteProps | undefined;
  if (_.isUndefined(notes)) {
    error = new DendronError({ msg: `no note found. ${hint}` });
    return { error, note };
  }
  if (notes.length > 1) {
    error = new DendronError({ msg: `multiple notes found for link: ${hint}` });
    return { error, note };
  }
  if (notes.length < 1) {
    error = new DendronError({
      msg: `no notes found for link: ${JSON.stringify(hint)}`,
    });
    return { error, note };
  }
  note = notes[0];
  return { error, note };
}

export class LinkUtils {
  static isAlias(link: string) {
    return link.indexOf("|") !== -1;
  }

  static hasFilter(link: string) {
    return link.indexOf(">") !== -1;
  }

  static parseAliasLink(link: string) {
    const [alias, value] = link.split("|").map(_.trim);
    return { alias, value: NoteUtils.normalizeFname(value) };
  }

  static parseDendronURI(linkString: string) {
    if (linkString.startsWith(CONSTANTS.DENDRON_DELIMETER)) {
      const [vaultName, link] = linkString
        .split(CONSTANTS.DENDRON_DELIMETER)[1]
        .split("/");
      return {
        vaultName,
        link,
      };
    }
    return {
      link: linkString,
    };
  }

  static parseLinkV2(linkString: string) {
    const LINK_NAME = "[^#\\|>]+";
    const re = new RegExp(
      "" +
        // alias?
        `(` +
        `(?<alias>${LINK_NAME}(?=\\|))` +
        "\\|" +
        ")?" +
        // name
        `(?<value>${LINK_NAME})` +
        // anchor?
        `(#(?<anchor>${LINK_NAME}))?` +
        // filters?
        `(>(?<filtersRaw>.*))?`,
      "i"
    );
    const out = linkString.match(re);
    if (out) {
      let { alias, value, anchor } = out.groups as any;
      let vaultName: string | undefined;
      ({ vaultName, link: value } = this.parseDendronURI(value));
      if (!alias) {
        alias = value;
      }
      alias = _.trim(alias);
      value = _.trim(value);
      return { alias, value, anchorHeader: anchor, vaultName };
    } else {
      return null;
    }
  }

  static parseLink(linkMatch: string) {
    linkMatch = NoteUtils.normalizeFname(linkMatch);
    let out: WikiLinkProps = {
      value: linkMatch,
      alias: linkMatch,
    };
    if (LinkUtils.isAlias(linkMatch)) {
      out = LinkUtils.parseAliasLink(linkMatch);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#").map(_.trim);
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!LinkUtils.isAlias(linkMatch)) {
        out.alias = value;
      }
    }
    return out;
  }
}

export class RemarkUtils {
  static oldNoteRef2NewNoteRef(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as DendronASTRoot;
        //@ts-ignore
        let notesRefLegacy: NoteRefNoteV4_LEGACY[] = selectAll("refLink", root);
        const slugger = getSlugger();
        notesRefLegacy.map((noteRefLegacy) => {
          // @ts-ignore;
          noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
          const { anchorStart, anchorEnd } = noteRefLegacy.data.link.data;
          if (anchorStart) {
            noteRefLegacy.data.link.data.anchorStart = normalizev2(
              anchorStart,
              slugger
            );
          }
          if (anchorEnd) {
            noteRefLegacy.data.link.data.anchorEnd = normalizev2(
              anchorEnd,
              slugger
            );
          }
        });
        changes.push({
          note,
          status: "update",
        });
        // if (out2.length > 0) {
        //   // const noteRefLegacy = root.children.splice(idx, 1)[0] as unknown as  NoteRefNoteV4_LEGACY;
        //   const noteRefLegacy = out2[0];
        //   // root.children[idx] as unknown as NoteRefNoteV4_LEGACY;
        //   // const noteRefNew: NoteRefNoteV4= {
        //   //   type: DendronASTTypes.REF_LINK_V2,
        //   //   data: {
        //   //     link: noteRefLegacy.data.link,
        //   //   },
        //   //   value: noteRefLegacy.value
        //   // }
        //   // @ts-ignore;
        //   noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
        //   // // @ts-ignore
        //   // root.children.splice(idx, 1, noteRefNew);
        //   changes.push({
        //     note,
        //     status: "update",
        //   });
        // }
      };
    };
  }

  static h1ToTitle(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children.splice(idx, 1)[0] as Heading;
          if (head.children.length === 1 && head.children[0].type === "text") {
            note.title = head.children[0].value;
          }
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }

  static h1ToH2(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children[idx] as Heading;
          head.depth = 2;
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }
}
