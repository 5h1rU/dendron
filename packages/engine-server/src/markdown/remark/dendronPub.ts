import { DendronError, NoteUtilsV2 } from "@dendronhq/common-all";
import _ from "lodash";
import { Image, Root } from "mdast";
import Unified, { Transformer } from "unified";
import { Node } from "unist";
import u from "unist-builder";
import visit from "unist-util-visit";
import { VFile } from "vfile";
import { DendronASTDest, NoteRefDataV4, WikiLinkNoteV4 } from "../types";
import { MDUtilsV4 } from "../utils";
import { convertNoteRefAST, NoteRefsOpts } from "./noteRefs";
import { addError, getNoteOrError } from "./utils";

type PluginOpts = NoteRefsOpts & {
  assetsPrefix?: string;
  insertTitle?: boolean;
};

function plugin(this: Unified.Processor, opts?: PluginOpts): Transformer {
  const proc = this;
  const { dest, vault, fname } = MDUtilsV4.getDendronData(proc);
  function transformer(tree: Node, _file: VFile) {
    let root = tree as Root;
    const { error, engine } = MDUtilsV4.getEngineFromProc(proc);

    if (opts?.insertTitle && root.children) {
      if (!fname || !vault) {
        throw new DendronError({
          msg: `no fname for node: ${JSON.stringify(tree)}`,
        });
      }
      const note = NoteUtilsV2.getNoteByFnameV4({
        fname,
        notes: engine.notes,
        vault: vault,
      });
      if (!note) {
        throw new DendronError({ msg: `no note found for ${fname}` });
      }
      const idx = _.findIndex(root.children, (ent) => ent.type !== "yaml");
      root.children.splice(
        idx,
        0,
        u("heading", { depth: 1 }, [u("text", note.title)])
      );
    }
    visit(tree, (node, _idx, parent) => {
      if (
        node.type === "wikiLink" &&
        dest !== DendronASTDest.MD_ENHANCED_PREVIEW
      ) {
        let _node = node as WikiLinkNoteV4;
        let value = node.value as string;
        const data = _node.data;
        if (error) {
          addError(proc, error);
        }

        const copts = opts?.wikiLinkOpts;
        if (copts?.useId) {
          // TODO: check for vault
          const notes = NoteUtilsV2.getNotesByFname({
            fname: value,
            notes: engine.notes,
            vault,
          });
          const { error, note } = getNoteOrError(notes, value);
          if (error) {
            addError(proc, error);
          } else {
            value = note!.id;
          }
        }
        const alias = data.alias ? data.alias : value;
        const href = `${copts?.prefix || ""}${value}.html${
          data.anchorHeader ? "#" + data.anchorHeader : ""
        }`;
        const exists = true;
        // for rehype
        //_node.value = newValue;
        _node.value = alias;
        _node.data = {
          alias,
          permalink: href,
          exists: exists,
          hName: "a",
          hProperties: {
            // className: classNames,
            href,
          },
          hChildren: [
            {
              type: "text",
              value: alias,
            },
          ],
        };
      }
      if (
        node.type === "refLink" &&
        dest !== DendronASTDest.MD_ENHANCED_PREVIEW
      ) {
        const ndata = node.data as NoteRefDataV4;
        const copts: NoteRefsOpts = {
          wikiLinkOpts: opts?.wikiLinkOpts,
          prettyRefs: opts?.prettyRefs,
        };
        const { data } = convertNoteRefAST({
          link: ndata.link,
          proc,
          compilerOpts: copts,
        });
        if (data) {
          parent!.children = data;
        }
      }
      if (node.type === "image" && dest === DendronASTDest.HTML) {
        let imageNode = node as Image;
        if (opts?.assetsPrefix) {
          imageNode.url = opts.assetsPrefix + imageNode.url;
        }
      }
    });
    return tree;
  }
  return transformer;
}

export { plugin as dendronPub };
export { PluginOpts as DendronPubOpts };
