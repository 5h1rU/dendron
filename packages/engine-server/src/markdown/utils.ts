import { DendronError, DEngineClientV2, DVault } from "@dendronhq/common-all";
import _ from "lodash";
import { Heading } from "mdast";
import { paragraph, root, text } from "mdast-builder";
// @ts-ignore
import katex from "rehype-katex";
import raw from "rehype-raw";
import rehypeStringify from "rehype-stringify";
import remark from "remark";
import frontmatterPlugin from "remark-frontmatter";
import math from "remark-math";
import remarkParse from "remark-parse";
import remark2rehype from "remark-rehype";
import { default as unified, default as Unified, Processor } from "unified";
import { Node, Parent } from "unist";
import { dendronPub, DendronPubOpts } from "./remark/dendronPub";
import { noteRefs, NoteRefsOpts } from "./remark/noteRefs";
import { wikiLinks, WikiLinksOpts } from "./remark/wikiLinks";
import { DendronASTData, DendronASTDest } from "./types";
// @ts-ignore
import highlight from "remark-highlight.js";

const toString = require("mdast-util-to-string");

type ProcOpts = {
  engine: DEngineClientV2;
};

export class MDUtilsV4 {
  static findIndex(array: Node[], fn: any) {
    for (var i = 0; i < array.length; i++) {
      if (fn(array[i], i)) {
        return i;
      }
    }
    return -1;
  }

  static genMDMsg(msg: string): Parent {
    return root(paragraph(text(msg)));
  }

  static getDendronData(proc: Processor) {
    return proc.data("dendron") as DendronASTData;
  }

  static setDendronData(proc: Processor, data: DendronASTData) {
    const _data = proc.data("dendron") as DendronASTData;
    return proc.data("dendron", { ..._data, ...data });
  }

  static getEngineFromProc(proc: Unified.Processor) {
    const engine = proc.data("engine") as DEngineClientV2;
    let error: DendronError | undefined;
    if (_.isUndefined(engine) || _.isNull(engine)) {
      error = new DendronError({ msg: "engine not defined" });
    }
    return {
      error,
      engine,
    };
  }

  static getNoteRefLvl(proc: Unified.Processor): number {
    return (proc.data("noteRefLvl") as number) || 0;
  }

  static setNoteRefLvl(proc: Unified.Processor, lvl: number) {
    return proc.data("noteRefLvl", lvl);
  }

  static isHeading(node: Node, text: string, depth?: number) {
    if (node.type !== "heading") {
      return false;
    }

    // wildcard is always true
    if (text === "*") {
      return true;
    }
    if (text) {
      var headingText = toString(node);
      return text.trim().toLowerCase() === headingText.trim().toLowerCase();
    }

    if (depth) {
      return (node as Heading).depth <= depth;
    }

    return true;
  }
  static proc(opts: ProcOpts) {
    const { engine } = opts;
    const errors: DendronError[] = [];
    let _proc = remark()
      .use(remarkParse, { gfm: true })
      .data("errors", errors)
      .data("engine", engine)
      .use(frontmatterPlugin, ["yaml"])
      .use({ settings: { listItemIndent: "1", fences: true, bullet: "-" } });
    return _proc;
  }

  static procFull(
    opts: ProcOpts & {
      dest: DendronASTDest;
      vault?: DVault;
      wikiLinksOpts?: WikiLinksOpts;
      noteRefOpts?: NoteRefsOpts;
      publishOpts?: DendronPubOpts;
      mathOpts?: {
        katex?: boolean;
      };
    }
  ) {
    const { dest, vault } = opts;
    let proc = this.proc(opts)
      .data("dendron", { dest, vault } as DendronASTData)
      .use(wikiLinks, opts.wikiLinksOpts)
      .use(noteRefs, { ...opts.noteRefOpts, wikiLinkOpts: opts.wikiLinksOpts });
    if (opts.mathOpts?.katex) {
      proc = proc.use(math);
    }
    // if (dest === DendronASTDest.HTML) {
    proc = proc.use(dendronPub, {
      ...opts.publishOpts,
      wikiLinkOpts: opts.wikiLinksOpts,
    });
    proc = proc.data("procFull", proc().freeze());
    return proc;
    //return proc;
  }

  static procRehype(opts: {
    proc?: Processor;
    mdPlugins?: Processor[];
    mathjax?: boolean;
  }) {
    const { proc, mdPlugins } = _.defaults(opts, { mdPlugins: [] });
    let _proc = proc || unified().use(remarkParse, { gfm: true });
    _proc = _proc.use(highlight);
    mdPlugins.forEach((p) => {
      _proc = _proc.use(p);
    });
    _proc = _proc.use(remark2rehype, { allowDangerousHtml: true }).use(raw);
    if (opts.mathjax) {
      _proc = _proc.use(katex);
    }
    return _proc.use(rehypeStringify);
  }
}
