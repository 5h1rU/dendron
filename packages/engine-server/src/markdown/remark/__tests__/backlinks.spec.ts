import { DEngineClientV2 } from "@dendronhq/common-all";
import {
  AssertUtils,
  ENGINE_HOOKS,
  NoteTestUtilsV4,
  runEngineTestV4,
} from "@dendronhq/common-test-utils";
import { createEngine } from "../../../enginev2";
import { DendronASTData, DendronASTDest } from "../../types";
import { MDUtilsV4 } from "../../utils";
import { DendronPubOpts } from "../dendronPub";
import { genDendronData } from "./utils";

// runs all the processes
function proc(
  engine: DEngineClientV2,
  dendron: DendronASTData,
  opts?: DendronPubOpts
) {
  return MDUtilsV4.procFull({
    fname: "PLACEHOLDER",
    engine,
    ...dendron,
    publishOpts: opts,
  });
}

describe("backlinks", () => {
  let dendronData = {
    dest: DendronASTDest.HTML,
  };

  test("basic", async () => {
    await runEngineTestV4(
      async ({ engine, vaults }) => {
        const vault = vaults[0];
        const resp = await MDUtilsV4.procRehype({
          proc: proc(
            engine,
            genDendronData({ ...dendronData, fname: "beta", vault })
          ),
        }).process("");
        // should be one backlink
        expect(resp).toMatchSnapshot();
        expect(
          await AssertUtils.assertInString({
            body: resp.contents as string,
            match: [`<a href="alpha.html">alpha (vault1)</a></p>`],
          })
        ).toBeTruthy();
      },
      { expect, createEngine, preSetupHook: ENGINE_HOOKS.setupLinks }
    );
  });

  test("multiple links", async () => {
    await runEngineTestV4(
      async ({ engine, vaults }) => {
        const vault = vaults[0];
        const resp = await MDUtilsV4.procRehype({
          proc: proc(
            engine,
            genDendronData({ ...dendronData, fname: "one", vault })
          ),
        }).process("");
        // should be one backlink
        expect(resp).toMatchSnapshot();
        expect(
          await AssertUtils.assertInString({
            body: resp.contents as string,
            match: [
              `<a href="three.html">three (vault1)</a>`,
              `<a href="two.html">two (vault1)</a>`,
            ],
          })
        ).toBeTruthy();
      },
      {
        expect,
        createEngine,
        preSetupHook: async (opts) => {
          const { wsRoot, vaults } = opts;
          const vault = vaults[0];
          await NoteTestUtilsV4.createNote({
            fname: "one",
            vault,
            wsRoot,
            body: "One body",
          });
          await NoteTestUtilsV4.createNote({
            fname: "two",
            vault,
            wsRoot,
            body: "[[one]]",
          });
          await NoteTestUtilsV4.createNote({
            fname: "three",
            vault,
            wsRoot,
            body: "[[one]]",
          });
        },
      }
    );
  });
});
