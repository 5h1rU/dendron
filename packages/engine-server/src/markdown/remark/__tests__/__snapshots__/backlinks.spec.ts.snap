// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backlinks basic 1`] = `
VFile {
  "contents": "<hr>
<h2 id=\\"backlinks\\"><a aria-hidden=\\"true\\" class=\\"anchor-heading\\" href=\\"#backlinks\\"><svg aria-hidden=\\"true\\" viewBox=\\"0 0 16 16\\"><use xlink:href=\\"#svg-link\\"></use></svg></a>Backlinks</h2>
<ul>
<li><a href=\\"alpha.html\\">alpha (vault1)</a></li>
<li><a href=\\"omega.html\\">omega (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`backlinks multiple links 1`] = `
VFile {
  "contents": "<hr>
<h2 id=\\"backlinks\\"><a aria-hidden=\\"true\\" class=\\"anchor-heading\\" href=\\"#backlinks\\"><svg aria-hidden=\\"true\\" viewBox=\\"0 0 16 16\\"><use xlink:href=\\"#svg-link\\"></use></svg></a>Backlinks</h2>
<ul>
<li><a href=\\"three.html\\">three (vault1)</a></li>
<li><a href=\\"two.html\\">two (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
